src/coreclr/jit/optcse.cpp

```diff
@@ -1826,168 +1833,1348 @@ bool CSE_HeuristicCommon::CanConsiderTree(GenTree* tree, bool isReturn)
+//
+// Notes:
+//   This converts preferences to likelihoods using softmax, and then
+//   randomly selects a candidate proportional to its likelihood.
+//
+void CSE_HeuristicRL::SoftmaxPolicy()
+{
+    if (m_verbose)
+    {
+        printf("RL using softmax policy\n");
+    }
+
+    // Number of choices is num candidates + 1, since
+    // early stopping is also a choice.
+    //
+    const int          numCandidates = m_pCompiler->optCSECandidateCount;
+    ArrayStack<Choice> choices(m_pCompiler->getAllocator(CMK_CSE), numCandidates + 1);
+    bool               first = true;
+
+    while (true)
+    {
+        Choice& choice = ChooseSoftmax(choices);
+
+        if (first)
+        {
+            for (int i = 0; i < choices.Height(); i++)
+            {
+                Choice& option = choices.TopRef(i);
+                if (option.m_dsc == nullptr)
+                {
+                    m_baseLikelihoods->push_back(0);
+                }
+                else
+                {
+                    m_baseLikelihoods->push_back(option.m_dsc->csdIndex);
+                }
+                m_baseLikelihoods->push_back(option.m_softmax);
+            }
+            first = false;
+        }
+
+        CSEdsc* const dsc = choice.m_dsc;
+
+        if (dsc == nullptr)
+        {
+            m_likelihoods->push_back(choice.m_softmax);
+            break;
+        }
+
+        // purge this CSE from sortTab so we won't choose it again
+        //
+        assert(sortTab[dsc->csdIndex - 1] == dsc);
+        sortTab[dsc->csdIndex - 1] = nullptr;
+
+        // ChooseCSE should only choose viable options
+        //
+        assert(dsc->IsViable());
+
+        CSE_Candidate candidate(this, dsc);
+
+        if (m_verbose)
+        {
+            printf("\nRL attempting " FMT_CSE "\n", candidate.CseIndex());
+        }
+
+        JITDUMP("CSE Expression : \n");
+        JITDUMPEXEC(m_pCompiler->gtDispTree(candidate.Expr()));
+        JITDUMP("\n");
+
+        PerformCSE(&candidate);
+        madeChanges = true;
+        m_likelihoods->push_back(choice.m_softmax);
+    }
+
+    return;
+}
+
+//------------------------------------------------------------------------
+// GetFeatures: extract features for this CSE
+//
+// Arguments:
+//    cse - cse descriptor, or nullptr for the option to stop doing CSEs.
+//    features - array to fill in with feature values
+//
+// Notes:
+//    Current set of features:
+//    0. cse costEx
+//    1. cse use count weighted (log)
+//    2. cse def count weighted (log)
+//    3. cse costSz
+//    4. cse use count
+//    5. cse def count
+//    6. cse live across call (0/1)
+//    7. cse is int (0/1)
+//    8. cse is a constant, but not shared (0/1)
+//    9. cse is a shared const (0/1)
+//   10. cse costEx is <= MIN_CSE_COST (0/1)
+//   11. cse is a constant and live across call (0/1)
+//   12. cse is a constant and min cost (0/1)
+//   13. cse is a constant and NOT min cost (0/1)
```

Comment:
Just wondering - are you going to take platform's features into account such as number of callee-saved regs (for GPR and floats)?

Reply:
Yes, we will need to add something like this -- right now the mechanisms to decide not to do a CSE are too weak.

I have follow-on changes that add some, but I'm not happy with them yet.